pyproject.toml

```
@@ -0,0 +1,3 @@
+[build-system]
+requires = ["setuptools >= 40.6.0", "numpy>=1.10", "cython>=0.28"]
```

This will fail if pip builds a wheel based on a numpy with a different ABI to the numpy installed due to install_requires

---

Sure, but does this need to be addressed by this PR? This sounds like more of an issue with building/distributing wheels in general and not really specific to this PR. Generalizing a bit, it sounds like your concern is that a wheel will be incompatible with future versions of one of the dependencies, numpy in this case. This problem is not specific to numpy, and could occur if any of cartopy's dependencies make a breaking change.

---

Yes, this is different because the whl created and cached will be ABI incompatible and so a simple reinstall won't fix it

---

I understand that numpy is both a build time requirement and a runtime requirement, which means that any given wheel may not work for all the versions of numpy that it claims to work with. 

Are you more concerned about the build process failing or a later user installation failing? 

I think the former is pretty unlikely. Numpy would have to issue an ABI breaking release between when the wheel is built and tested. As for the latter, I still don't see how this is specific to numpy since no build artifact (even an sdist) can ensure forward compatibility against all future releases of dependencies. For instance, the [xarray v0.16.1  sdist on pypi](https://files.pythonhosted.org/packages/70/54/d3416261ec3e427996f196a1a5489c553d4a4624eb61932d090f6d7c7680/xarray-0.16.1.tar.gz) claims to work with any pandas > 0.25. This is a lie, it doesn't work with pandas 1.0.0. This is why pinning package versions is so useful for application developers.

---

When you install a package from pypi it builds the wheel and caches it locally. If you install the same package and version in another virtual environment with a numpy with a different ABI it will install but not run correctly

---

I understand that, but doesn't this problem exists for the current master? How is this specific to this PR?

---

With this PR it's more likely to occur as the wheel will only compile with the latest numpy and make it impossible to use with older versions

---

Here's some more details: https://github.com/numpy/numpy/issues/5888#issuecomment-313283876

---

Also, the spirit of this PR is to facilitate using external tooling (e.g. poetry) that would explicitly allow you to avoid this potentially inevitable situation when using floating releases. The issue is not specific to numpy ABI-related incompatibilities, it could occur for any type of breaking change (e.g. pure python API breaking change). I think the issue you are raising is that a source distribution (sdist) could have better forward compatibility than a compiled artifact (wheel). I agree, but this is again an issue with wheels in general, and not this PR.

---

> it could occur for any type of breaking change (e.g. pure python API breaking change).

No it cannot. This is specific to numpy ABI compatibility

---

https://github.com/numpy/numpy/issues/5888#issuecomment-503563618

> If I recall correctly, SciPy pins its build system to the oldest numpy it wishes to support, but tests against later numpy versions.

cartopy should copy scipy in this regard

---

```suggestion
requires = ["setuptools >= 40.6.0", "numpy==1.10", "cython>=0.28"]
```

---

or even better the correct version for each py version:

https://github.com/scipy/scipy/blob/master/pyproject.toml#L6

---

```suggestion
requires = [
    "setuptools >= 40.6.0",
    "cython>=0.28",
    "numpy==1.16.5; python_version=='3.6'",
    "numpy==1.16.5; python_version=='3.7'",
    "numpy==1.17.3; python_version=='3.8'",
    # do not pin numpy on future versions of python to avoid incompatible numpy and python versions
    "numpy; python_version>='3.9'",
]
```

---

see also NEP 29: https://github.com/SciTools/cartopy/issues/1518

---

> cartopy should copy scipy in this regard

Sounds good to me. It sounds like strategy will ensure the widest possible range of compatibility.

---

```suggestion
requires = [
    "wheel",
    "setuptools",
    "Cython>=0.29.18",
    "numpy==1.16.5; python_version=='3.7'",
    "numpy==1.17.3; python_version=='3.8'",
    # do not pin numpy on future versions of python to avoid incompatible numpy and python versions
    "numpy; python_version>='3.9'",
]
```

if you want to follow NEP29